src/audio/component.c

```diff
@@ -496,6 +496,62 @@ void audio_stream_copy_to_linear(const struct audio_stream *source, int ioffset,
 	}
 }
 
+static inline bool comp_check_eos(struct comp_dev *dev)
+{
+	enum sof_audio_stream_state sink_state = STREAM_STATE_INITIAL;
+	struct comp_buffer *buffer;
+
+	if (!dev->pipeline->expect_eos)
+		return false;
+
+	comp_dev_for_each_producer(dev, buffer) {
+		struct sof_source *source = audio_buffer_get_source(&buffer->audio_buffer);
+		enum sof_audio_stream_state state = source_get_state(source);
+
+		if (source_get_pipeline_id(source) != dev->pipeline->pipeline_id)
+			continue;
+
+		if (state == STREAM_STATE_END_OF_STREAM_FLUSH) {
+			/* Earlier in the pipeline, there is a DP module that has reached
+			 * the EOS state. However, silence is generated to flush its internal
+			 * buffers, so pass this state to the output buffers.
+			 */
+			comp_dbg(dev, "comp_check_eos() - EOS flush detected");
+			sink_state = STREAM_STATE_END_OF_STREAM_FLUSH;
+			break;
+		} else if (state == STREAM_STATE_END_OF_STREAM) {
+			/* EOS is detected, so we need to set the sink state to STREAM_STATE_EOS. */
+			size_t min_avail = source_get_min_available(source);
+
+			if (source_get_data_available(source) < min_avail) {
+				comp_dbg(dev, "comp_check_eos() - EOS detected");
+				if (dev->ipc_config.proc_domain == COMP_PROCESSING_DOMAIN_DP) {
+					/* For DP modules, fill missing input data with silence to
+					 * allow it to process the remaining data.
```

Comment:
What will happen on output? When DP finishes processing, it will load data to output buffer(s). What if the output buffer is in EOS state at the moment?

Reply:
After detecting EOS, DP modules receive silence at input so that they can process internally buffered data.